### PR TITLE
Feature/condo/sberdoma 507/ticket comment schema

### PR DESCRIPTION
**Pull request by AntonAL**
TicketComment schema with access rights and tests

### CODE REVIEW

**Review by dkoviazin — APPROVED** [review]
Ok

File: apps/condo/domains/ticket/gql.js [hunk 1]
```diff
@@ -111,6 +111,9 @@ const TicketChange = generateGqlQueries('TicketChange', TICKET_CHANGE_FIELDS)
 const TICKET_FILE_FIELDS = `{ id file { id originalFilename publicUrl mimetype } organization { id } ticket { id } ${COMMON_FIELDS} }`
 const TicketFile = generateGqlQueries('TicketFile', TICKET_FILE_FIELDS)
 
+const TICKET_COMMENT_FIELDS = `{ ticket { id } user { id } content ${COMMON_FIELDS} }`
```
Comment (pahaz) [inline apps/condo/domains/ticket/gql.js:114]:
```suggestion
const TICKET_COMMENT_FIELDS = `{ ticket { id } user { id name } content ${COMMON_FIELDS} }`
```

File: apps/condo/domains/ticket/schema/TicketComment.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+        // This field will take part in business logic and should be declared here explicitly.
+        // That's why `createdBy` field, generated by `tracked` plugin, is not used for association of a comment with ticket.
+        user: {
+            schemaDoc: 'User, who created the comment',
+            type: Relationship,
+            ref: 'User',
+            isRequired: true,
+            knexOptions: { isNotNullable: true }, // Required relationship only!
+            kmigratorOptions: { null: false, on_delete: 'models.CASCADE' },
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketComment.js:35]: We should check that you have access to set the user to non-self value. To add TODO for that

**Review by toplenboren — APPROVED** [review]
Approve

File: apps/condo/domains/ticket/schema/TicketComment.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        },
+
+        content: {
+            schemaDoc: 'Plain text content',
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketComment.js:39]: it's better to add some simple normalization. UTF normalization + cleaning of redundant spaces. 
May be add a todo here and inside a ticket

File: apps/condo/domains/ticket/schema/TicketComment.test.js [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+const { createTestProperty } = require('@condo/domains/property/utils/testSchema')
+const { createTestOrganization } = require('@condo/domains/organization/utils/testSchema')
+const { createTestTicket } = require('@condo/domains/ticket/utils/testSchema')
+const { makeClientWithProperty } = require('@condo/domains/property/utils/testSchema')
+const { makeLoggedInAdminClient, makeClient, UUID_RE, DATETIME_RE } = require('@core/keystone/test.utils')
+const { TicketComment, createTestTicketComment, updateTestTicketComment } = require('@condo/domains/ticket/utils/testSchema')
+const { expectToThrowAccessDeniedErrorToObj, expectToThrowAccessDeniedErrorToObjects } = require('../../common/utils/testSchema')
+
+describe('TicketComment', () => {
+    describe('create', () => {
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketComment.test.js:18]: could you please change the default test template %) it's will help to others

File: apps/condo/domains/ticket/schema/TicketComment.test.js [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+            expect(obj.dv).toEqual(1)
+            expect(obj.sender).toEqual(attrs.sender)
+            expect(obj.v).toEqual(1)
+            expect(obj.newId).toEqual(null)
+            expect(obj.deletedAt).toEqual(null)
+            expect(obj.createdBy).toEqual(expect.objectContaining({ id: userClient.user.id }))
+            expect(obj.updatedBy).toEqual(expect.objectContaining({ id: userClient.user.id }))
+            expect(obj.createdAt).toMatch(DATETIME_RE)
+            expect(obj.updatedAt).toMatch(DATETIME_RE)
+            expect(obj.content).toMatch(attrs.content)
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketComment.test.js:43]: what about user field?

File: apps/condo/domains/ticket/schema/TicketComment.js (hunk 2 above)
Comment (pahaz) [reply apps/condo/domains/ticket/schema/TicketComment.js:35]: Or restrict to set non-self value if the user is not an admin ...

**Review by pahaz — APPROVED** [review]

File: apps/condo/domains/ticket/schema/TicketComment.js (hunk 3 above)
Comment (AntonAL) [reply apps/condo/domains/ticket/schema/TicketComment.js:39]: https://doma.atlassian.net/browse/SBERDOMA-539
For juns

File: apps/condo/domains/ticket/schema/TicketComment.test.js (hunk 4 above)
Comment (AntonAL) [reply apps/condo/domains/ticket/schema/TicketComment.test.js:18]: https://doma.atlassian.net/browse/SBERDOMA-540

File: apps/condo/domains/ticket/schema/TicketComment.test.js (hunk 5 above)
Comment (AntonAL) [reply apps/condo/domains/ticket/schema/TicketComment.test.js:43]: Added in `ebc5cb7e`

File: apps/condo/domains/ticket/gql.js (hunk 1 above)
Comment (AntonAL) [reply apps/condo/domains/ticket/gql.js]: Added in `d35c80bd`

File: apps/condo/domains/ticket/schema/TicketComment.js (hunk 2 above)
Comment (AntonAL) [reply apps/condo/domains/ticket/schema/TicketComment.js:35]: Added in `62c88c73`

File: apps/condo/domains/ticket/access/TicketComment.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    return false
+}
+
+async function canSetUserField ({ authentication: { item: user }, originalInput, addFieldValidationError }) {
+    if (user.isAdmin) return true
+    if (get(originalInput, ['user', 'connect', 'id']) !== user.id) {
+        return false
+    }
+    return true
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/TicketComment.js:53]: Wow, it's better to false in other cases!

**Pull request merged**